File: Android/app/src/main/java/org/mobiletriage/localverify/core/InputLimits.kt
package org.mobiletriage.localverify.core

object InputLimits {
    fun read(input: java.io.InputStream, maximum: Int): ByteArray {
        val out = java.io.ByteArrayOutputStream()
        val buffer = ByteArray(65536)
        while (true) {
            val count = input.read(buffer, 0, minOf(buffer.size, maximum - out.size() + 1))
            if (count < 0) break
            require(count <= maximum - out.size()) { "Input exceeds byte limit" }
            out.write(buffer, 0, count)
        }
        return out.toByteArray()
    }
    const val INDICATOR_BYTES = 5 * 1024 * 1024
    fun text(text: String) {
        require(text.length <= 16 * 1024 * 1024) { "Text exceeds size limit" }
        var lines = 1
        var length = 0
        for (char in text) {
            if (char == '\n' || char == '\r') { lines++; length = 0 } else length++
            require(lines <= 500_000 && length <= 1024 * 1024) { "Text line limit reached" }
        }
    }
    // Conservative preflight before recursive platform JSON parsing, not validation.
    fun json(text: String) {
        var depth = 0
        var tokens = 0
        var quote: Char? = null
        var escaped = false
        for ((index, char) in text.withIndex()) {
            if (index % 65536 == 0 && Thread.currentThread().isInterrupted) {
                throw java.util.concurrent.CancellationException("Analysis interrupted")
            }
            if (quote != null) {
                if (escaped) escaped = false
                else if (char == '\\') escaped = true
                else if (char == quote) quote = null
            } else {
                if (char == '"' || char == '\'') { quote = char; tokens++ }
                if (char == '{' || char == '[') { depth++; tokens++ }
                if (char == '}' || char == ']') depth = maxOf(0, depth - 1)
                if (char == ',') tokens++
                require(depth <= 64 && tokens <= 200_000) { "JSON complexity limit reached" }
            }
        }
    }
    fun indicators(values: List<Indicator>) {
        require(values.size <= 10_000 && values.all {
            it.value.isNotEmpty() && it.value.length <= 2048 && it.id.length <= 1024 &&
                (it.campaigns?.size ?: 0) <= 16 && it.campaigns.orEmpty().all { label -> label.length <= 128 }
        }) { "Indicator size/count limit reached" }
    }
}

File: Sources/TriageCore/InputLimits.swift
import Foundation

/// Budgets are checked before Foundation parsing or allocation-heavy indexing.
public enum InputLimits {
    public enum ResourceLimit: LocalizedError {
        case exceeded(String)
        public var errorDescription: String? {
            switch self { case .exceeded(let message): return message }
        }
    }
    public static let indicatorBytes = 5 * 1024 * 1024
    public static let textBytes = 16 * 1024 * 1024
    public static func read(_ url: URL, maximum: Int) throws -> Data {
        let file = try FileHandle(forReadingFrom: url)
        defer { try? file.close() }
        var data = Data()
        while let chunk = try file.read(upToCount: min(65536, maximum - data.count + 1)), !chunk.isEmpty {
            try Task.checkCancellation()
            guard chunk.count <= maximum - data.count else { throw TriageError.invalid("Input exceeds byte limit") }
            data.append(chunk)
        }
        return data
    }
    public static func text(_ text: String) throws {
        guard text.utf8.count <= textBytes else { throw ResourceLimit.exceeded("Text exceeds byte limit") }
        var lines = 1; var length = 0
        for byte in text.utf8 {
            if byte == 10 || byte == 13 { lines += 1; length = 0 } else { length += 1 }
            guard lines <= 500_000, length <= 1024 * 1024 else { throw ResourceLimit.exceeded("Text line limit reached") }
        }
    }
    /// A conservative lexical budget, not a JSON validator. Run before JSONSerialization.
    public static func json(_ data: Data) throws {
        var depth = 0; var tokens = 0; var quoted = false; var escaped = false
        for (index, byte) in data.enumerated() {
            if index % 65536 == 0 { try Task.checkCancellation() }
            if quoted {
                if escaped { escaped = false }
                else if byte == 92 { escaped = true }
                else if byte == 34 { quoted = false }
            } else {
                if byte == 34 { quoted = true; tokens += 1 }
                if byte == 123 || byte == 91 { depth += 1; tokens += 1 }
                if byte == 125 || byte == 93 { depth = max(0, depth - 1) }
                if byte == 44 { tokens += 1 }
                guard depth <= 64, tokens <= 200_000 else { throw ResourceLimit.exceeded("JSON complexity limit reached") }
            }
        }
    }
    public static func indicators(_ values: [Indicator]) throws {
        guard values.count <= 10_000, values.allSatisfy({ !$0.value.isEmpty && $0.value.utf8.count <= 8192 && $0.id.utf8.count <= 1024 && ($0.campaigns ?? []).count <= 16 && ($0.campaigns ?? []).allSatisfy { $0.utf8.count <= 128 } }) else {
            throw TriageError.invalid("Indicator size/count limit reached")
        }
    }
}
